README.txt

    OpenX - The ad server for publishers who are serious about making money
    ------------------------------------------------------------------------

    OpenX is an open source ad serving tool that enables publishers to:

      * Serve ads on their websites;

      * Manage their campaigns from different advertisers and/or ad networks
          using the simple, easy-to-use interface;

      * Track and report on campaign success, including click-through
          rates;

      * Set rules to target the delivery of campaigns, or even ads, to
          specific users, to help maximise the effectiveness of campaigns.

    OpenX was formerly known as Openads and phpAdsNew.



    Version 2.8.7
    Copyright (c) 2003-2009 OpenX Limited
    ------------------------------------------------------------------------

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

    ------------------------------------------------------------------------

    In addition, as a special exception, TOMO <groove@spencernetwork.org>
    gives permission to link the code of this program with the FTP library
    (or with modified versions of LIB-FTP that use the same license as LIB-FTP),
    and distribute linked combinations including the two.  You must obey
    the GNU General Public License in all respects for all of the code used
    other than LIB-FTP.  If you modify this file, you may extend this exception
    to your version of the file, but you are not obligated to do so.  If
    you do not wish to do so, delete this exception statement from your version.



    Support:
    ------------------------------------------------------------------------
    Questions about OpenX should be asked on the forum at
    http://forum.openx.org/.

    More information about can be found on the OpenX homepage at
    http://www.openx.org/.

    Bug reports and questions related to the development of OpenX are
    also welcome. Plese visit http://developer.openx.org.



    Download:
    ------------------------------------------------------------------------
    You can get the latest version from http://www.openx.org/.



    Credits:
    ------------------------------------------------------------------------
    Thanks to Marcus Baker for allowing OpenX Limited to distribute Simple
      Test for PHP along with OpenX. Simple Test is distributed under the
      Open Group Test Suite License. Please see the Simple Test LICENSE file
      in the lib/simpletest/ directory for details.

    This product includes the phpSniff library, which is distributed under
      the GNU Lesser General Public License. Please see the the phpSniff
      LICENSE file in the lib/phpSniff/ directory for details.

    This product includes GeoIP data created by MaxMind, available from
      http://maxmind.com/.

    FTP library: TOMO <groove@spencernetwork.org>



    http://www.openx.org/


RELEASE_NOTES.txt
+---------------------------------------------------------------------------+
| OpenX v2.8                                                                |
| =========                                                                 |
|                                                                           |
| Copyright (c) 2003-2009 OpenX Limited                                     |
| For contact details, see: http://www.openx.org/                           |
|                                                                           |
| This program is free software; you can redistribute it and/or modify      |
| it under the terms of the GNU General Public License as published by      |
| the Free Software Foundation; either version 2 of the License, or         |
| (at your option) any later version.                                       |
|                                                                           |
| This program is distributed in the hope that it will be useful,           |
| but WITHOUT ANY WARRANTY; without even the implied warranty of            |
| MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             |
| GNU General Public License for more details.                              |
|                                                                           |
| You should have received a copy of the GNU General Public License         |
| along with this program; if not, write to the Free Software               |
| Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA |
+---------------------------------------------------------------------------+

Release Notes: OpenX 2.8.7
============================

OpenX 2.8 is the next generation release of the hugely popular OpenX
adserver.

These release notes cover what's new, system requirements, download and
installation instructions, known issues and frequently asked questions for
OpenX 2.8.7. Please read these notes before reporting any bugs.

Please tell us what you think on the OpenX forum [http://forum.openx.org/].


What's New in OpenX 2.8.7
---------------------------

 * Release date: 2010-09-14

 * For complete details, see [https://developer.openx.org/].


System Requirements
-------------------

Before installing, please make sure your server meets the system
requirements [http://www.openx.org/help/2.8/requirements].


Downloading and Installing
--------------------------

 * Downloading OpenX

   You can always find the latest version of OpenX here
   [http://www.openx.org/download].

   Older and maintenance releases can be found here
   [http://www.openx.org/support/release-archive].

 * Installing OpenX 2.8.7

   Installing OpenX 2.8.7 is a straightforward process. Follow these steps
   [http://www.openx.org/help/2.8/install] and use the links to refer to
   detailed information on each step where necessary.

 * Upgrading OpenX 2.8.7

   Upgrading to OpenX 2.8.7 is a straightforward process. Follow these steps
   [http://www.openx.org/help/2.8/install] and use the links to refer to
   detailed information on each step where necessary.

   Please note: The upgrade process will upgrade your OpenX database,
   and the database may no longer be compatible with your old code base.
   Therefore, you should create a backup of your database, just in case.

   The upgrade wizard supports upgrading from the following previous
   releases of OpenX:

   ---------------------------------------------------------------
   | Product                             | Version(s)            |
   ---------------------------------------------------------------
   | OpenX 2.7 beta                      | All previous versions |
   ---------------------------------------------------------------
   | OpenX 2.6                           | All versions          |
   ---------------------------------------------------------------
   | OpenX 2.4                           | All versions          |
   ---------------------------------------------------------------
   | Openads 2.4                         | All versions          |
   ---------------------------------------------------------------
   | Openads 2.0 for MySQL               | 2.0.11-pr1            |
   |   (formerly phpAdsNew)              |                       |
   ---------------------------------------------------------------
   | Openads 2.3 alpha                   | 2.3.31-alpha-pr3      |
   |   (formerly Max Media Manager v0.3) |   (v0.3.31-alpha-pr3) |
   ---------------------------------------------------------------
   | Openads 2.3 beta                    | All versions          |
   ---------------------------------------------------------------
   | Max Media Manager v0.1              | v0.1.29-rc            |
   ---------------------------------------------------------------
   | Openads 2.0 for PostgreSQL          | 2.0.11-pr1            |
   |  (formerly phpPgAds)                |                       |
   ---------------------------------------------------------------

   If you are not running one of these versions, you will need to upgrade
   your existing installation before you will be able to upgrade to
   OpenX 2.8.7.

 * Uninstalling OpenX 2.8.7

   To remove OpenX from your server delete the installed files and
   database tables.


Known Issues
------------

This list covers some of the known problems with OpenX 2.8.7. Please
read this before reporting any new bugs.

 * The upgrade process may time out with very large databases. See the FAQ
   [http://www.openx.org/help/2.8/faq/large-db] for more information if you
   have a large database.

 * OpenX 2.8 requires PHP 5. If you are using PHP 4, you will need to
   update your web server to use PHP 5 before upgrading.

 * If you want to run OpenX with MySQL, please note that OpenX 2.8 requires
   MySQL 4.1 or higher. If you are using an older version of MySQL, you will
   need to update you database software before upgrading.


lib/OA.php
<?php

/*
+---------------------------------------------------------------------------+
| OpenX v2.8                                                                |
| ==========                                                                |
|                                                                           |
| Copyright (c) 2003-2009 OpenX Limited                                     |
| For contact details, see: http://www.openx.org/                           |
|                                                                           |
| This program is free software; you can redistribute it and/or modify      |
| it under the terms of the GNU General Public License as published by      |
| the Free Software Foundation; either version 2 of the License, or         |
| (at your option) any later version.                                       |
|                                                                           |
| This program is distributed in the hope that it will be useful,           |
| but WITHOUT ANY WARRANTY; without even the implied warranty of            |
| MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             |
| GNU General Public License for more details.                              |
|                                                                           |
| You should have received a copy of the GNU General Public License         |
| along with this program; if not, write to the Free Software               |
| Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA |
+---------------------------------------------------------------------------+
$Id: OA.php 62345 2010-09-14 21:16:38Z chris.nutting $
*/

###START_STRIP_DELIVERY
require_once MAX_PATH . '/lib/pear/Log.php';
require_once MAX_PATH . '/lib/pear/PEAR.php';
require_once MAX_PATH . '/lib/OX/Admin/Timezones.php';

/**
 * this is a method to capture select queries and write them to a logfile
 * for analysis purposes
 * to trigger set $GLOBALS['_MAX']['CONF']['debug']['logSQL'] = 1
 *
 * @param mdb2 connecction $oDbh
 */
function logSQL($oDbh, $scope, $message, $context)
{
    // don't log 'explain' queries or we spiral out of control
    // don't log queries against temporary tables (cos the tables won't exist to use for explain)
    if ((substr_count($message, 'EXPLAIN')==0) && (substr_count($message, 'tmp_')==0))
    {
        $log = fopen(MAX_PATH."/var/sql.log", 'a');

        $aStatements = $oDbh->options['log_statements'];

        foreach ($aStatements AS $statement)
        {
            $i = strpos($message, strtoupper($statement));
            if ($i > -1)
            {
                $query = $message;
                if ($i > 0)
                {
                    if (strpos($message,'PREPARE MDB2_STATEMENT')<0)
                    {
                        $query = substr($query,$i, strlen($query)-1);
                    }
                }
                $query = preg_replace('/[\s\t\n]+/',' ',$query);
                $query = str_replace('\n','',$query);
                $query = stripslashes($query);
                fwrite($log, "[".trim($scope)."] ".trim($query)."; \n");
            }
        }
        fclose($log);
    }
}

/**
 * The core OpenX class, providing handy methods that are useful everywhere!
 *
 * @package    OpenX
 * @author     Andrew Hill <andrew.hill@openx.org>
 */
class OA
{

    /**
     * A method to log debugging messages to the location configured by the user.
     *
     * Note: If the global variable $currentTimezone is set, where the array
     * is the result of OX_Admin_Timezones::getTimezone(), called BEFORE any
     * timezone information has been set (i.e. before the init script has been
     * called), then this method will ensure that all debug messages are logged
     * in the SERVER TIME ZONE, rather than the time zone that the software
     * happens to be running in (i.e. the current manager timezone, or UTC for
     * maintenance).
     *
     * @static
     * @param mixed $message     Either a string or a PEAR_Error object.
     * @param integer $priority  The priority of the message. One of:
     *                           PEAR_LOG_EMERG, PEAR_LOG_ALERT, PEAR_LOG_CRIT
     *                           PEAR_LOG_ERR, PEAR_LOG_WARNING, PEAR_LOG_NOTICE
     *                           PEAR_LOG_INFO, PEAR_LOG_DEBUG
     * @return boolean           True on success or false on failure.
     *
     * @TODO Logging to anything other than a file is probably broken - test!
     */
    function debug($message = null, $priority = PEAR_LOG_INFO)
    {
        $aConf = $GLOBALS['_MAX']['CONF'];
        global $tempDebugPrefix;
        // Logging is not activated
        if ($aConf['log']['enabled'] == false) {
            unset($GLOBALS['tempDebugPrefix']);
            return true;
        }
        // Is this a "no message" log?
        if (is_null($message) && $aConf['log']['type'] == 'file') {
            // Set the priority to the highest level, so it is always logged
            $priority = PEAR_LOG_EMERG;
        }
        // Deal with the config file containing the log level by
        // name or by number
        $priorityLevel = is_numeric($aConf['log']['priority']) ? $aConf['log']['priority'] :
            @constant($aConf['log']['priority']);
        if (is_null($priorityLevel)) {
            $priorityLevel = $aConf['log']['priority'];
        }
        if ($priority > $priorityLevel) {
            unset($GLOBALS['tempDebugPrefix']);
            return true;
        }
        // Grab DSN if we are logging to a database
        $dsn = ($aConf['log']['type'] == 'sql') ? Base::getDsn() : '';
        // Instantiate a logger object based on logging options
        $aLoggerConf = array(
            $aConf['log']['paramsUsername'],
            $aConf['log']['paramsPassword'],
            'dsn'        => $dsn,
            'mode'       => octdec($aConf['log']['fileMode']),
            'timeFormat' => '%b %d %H:%M:%S %z'
        );
        if (is_null($message) && $aConf['log']['type'] == 'file') {
            $aLoggerConf['lineFormat'] = '%4$s';
        } else if ($aConf['log']['type'] == 'file') {
            $aLoggerConf['lineFormat'] = '%1$s %2$s [%3$9s]  %4$s';
        }
        $ident = (!empty($GLOBALS['_MAX']['LOG_IDENT'])) ? $GLOBALS['_MAX']['LOG_IDENT'] : $aConf['log']['ident'];
        if (($ident == $aConf['log']['ident'] . '-delivery') && empty($aConf['deliveryLog']['enabled'])) {
            unset($GLOBALS['tempDebugPrefix']);
            return true;
        }
        if ($ident == $aConf['log']['ident'] . '-delivery') {
            $logFile = $aConf['deliveryLog']['name'];
            list($micro_seconds, $seconds) = explode(" ", microtime());
            $message = (round(1000 *((float)$micro_seconds + (float)$seconds))) - $GLOBALS['_MAX']['NOW_ms'] . 'ms ' . $message; 
        } else {
            $logFile = $aConf['log']['name'];
        }

        $ident .= (!empty($GLOBALS['_MAX']['thread_id'])) ? '-' . $GLOBALS['_MAX']['thread_id'] : '';

        $oLogger = &Log::singleton(
            $aConf['log']['type'],
            MAX_PATH . '/var/' . $logFile,
            $ident,
            $aLoggerConf
        );
        // If log message is an error object, extract info
        if (PEAR::isError($message)) {
            $userinfo = $message->getUserInfo();
            $message = $message->getMessage();
            if (!empty($userinfo)) {
                if (is_array($userinfo)) {
                    $userinfo = implode(', ', $userinfo);
                }
            $message .= ' : ' . $userinfo;
            }
        }
        // Obtain backtrace information
        $aBacktrace = debug_backtrace();
        if ($aConf['log']['methodNames']) {
            // Show from four calls up the stack, to avoid the
            // showing the PEAR error call info itself
            $aErrorBacktrace = $aBacktrace[4];
            if (isset($aErrorBacktrace['class']) && $aErrorBacktrace['type'] && isset($aErrorBacktrace['function'])) {
                $callInfo = $aErrorBacktrace['class'] . $aErrorBacktrace['type'] . $aErrorBacktrace['function'] . ': ';
                $message = $callInfo . $message;
            }
        }
        // Show entire stack, line-by-line
        if ($aConf['log']['lineNumbers']) {
            foreach($aBacktrace as $aErrorBacktrace) {
                if (isset($aErrorBacktrace['file']) && isset($aErrorBacktrace['line'])) {
                    $message .=  "\n" . str_repeat(' ', 20 + strlen($aConf['log']['ident']) + strlen($oLogger->priorityToString($priority)));
                    $message .= 'on line ' . $aErrorBacktrace['line'] . ' of "' . $aErrorBacktrace['file'] . '"';
                }
            }
        }
        // Log messages in the local server timezone, if possible
        global $serverTimezone;
        if (!empty($serverTimezone)) {
            $currentTimezone = OX_Admin_Timezones::getTimezone();
            OA_setTimeZone($serverTimezone);
        }
        // Log the message
        if (is_null($message) && $aConf['log']['type'] == 'file') {
            $message = ' ';
        } else if (!is_null($tempDebugPrefix) && $aConf['log']['type'] == 'file') {
            $message = $tempDebugPrefix . $message;
        }
        $result = $oLogger->log($message, $priority);
        // Restore the timezone
        if (!empty($currentTimezone)) {
            OA_setTimeZone($currentTimezone);
        }
        unset($GLOBALS['tempDebugPrefix']);
        return $result;
    }

    function switchLogIdent($name = 'debug')
    {
        if ($name == 'debug') {
            $GLOBALS['_MAX']['LOG_IDENT'] = $GLOBALS['_MAX']['CONF']['log']['ident'];
        } else {
            $GLOBALS['_MAX']['LOG_IDENT'] = $GLOBALS['_MAX']['CONF']['log']['ident'] . '-' . $name;
        }
    }

    /**
     * A method to temporarily set the debug message prefix string. The string
     * is un-set when debug() is called.
     *
     * @param string $prefix The prefix to add to a message logged when the
     *                       debug() method is next called, in the event that
     *                       the logging is to a file.
     */
    function setTempDebugPrefix($prefix)
    {
        global $tempDebugPrefix;
        $tempDebugPrefix = $prefix;
    }


    function logMem($msg='', $peak=false)
    {
        /*if (isset($aConf['debug']['logmem']) && $aConf['debug']['logmem'])
        {*/
            $aConf = $GLOBALS['_MAX']['CONF'];
            $oLogger = &Log::singleton(
                $aConf['log']['type'],
                MAX_PATH . '/var/memory.log',
                $aConf['log']['ident'],
                array()
            );
            $pid = getmypid();
            //$msg.= ' MEMORY USAGE (% KB PID ): ' . `ps --pid $pid --no-headers -o%mem,rss,pid`;
            $mem = `ps --pid $pid --no-headers -orss`;
            $mem = round((memory_get_usage()/1048576),2). ' / ps '.$mem;
            $msg = '['.rtrim($mem,chr(10)).']('.$msg.')';
            $aLast = array_pop(debug_backtrace());
            if ($aLast['function']=='logMem')
            {
                $msg.= str_replace(MAX_PATH,'',$aLast['file'].' -> line '.$aLast['line']);
            }
            else
            {
                $msg.= $aLast['class'] . $aLast['type'] . $aLast['function'] . ': ';
            }
            $oLogger->log($msg, PEAR_LOG_INFO);
            if ($peak)
            {
                $peak = memory_get_peak_usage()/1048576;
                $oLogger->log('PEAK: '.$peak, PEAR_LOG_INFO);
            }
        //}
    }

    function logMemPeak($msg='')
    {
        OA::logMem($msg, true);
    }

    /**
     * A method to obtain the current date/time, offset if required by the
     * user configured timezone.
     *
     * @static
     * @param string $format A PHP date() compatible formatting string, if
     *                       required. Default is "Y-m-d H:i:s".
     * @return string An appropriately formatted date/time string, representing
     *                the "current" date/time, offset if required.
     */
    function getNow($format = null)
    {
        if (is_null($format)) {
            $format = 'Y-m-d H:i:s';
        }
        return date($format);
    }

    /**
     * A method to obtain the current date/time in UTC
     *
     * @static
     * @param string $format A PHP date() compatible formatting string, if
     *                       required. Default is "Y-m-d H:i:s".
     * @return string An appropriately formatted date/time string, representing
     *                the "current" date/time, offset if required.
     */
    function getNowUTC($format = null)
    {
        if (is_null($format)) {
            $format = 'Y-m-d H:i:s';
        }
        return gmdate($format);
    }

    /**
     * A method to detect the available SSL enabling extensions
     *
     * @return mixed An array of the available extensions, or false if none is present
     */
    function getAvailableSSLExtensions()
    {
        $aResult = array();

        if (extension_loaded('curl')) {
            $aCurl = curl_version();
            if (!empty($aCurl['ssl_version'])) {
                $aResult[] = 'curl';
            }
        }
        if (extension_loaded('openssl')) {
            $aResult[] = 'openssl';
        }

        return count($aResult) ? $aResult : false;
    }

    /**
     * A method to strip unwanted ending tags from an OpenX version string.
     *
     * @static
     * @param string $version The original version string.
     * @param array  $aAllow  An array of allowed tags
     * @return string The stripped version string.
     */
    function stripVersion($version, $aAllow = null)
    {
        $allow = is_null($aAllow) ? '' : '|'.join('|', $aAllow);
        return preg_replace('/^v?(\d+.\d+.\d+(?:-(?:beta(?:-rc\d+)?|rc\d+'.$allow.'))?).*$/i', '$1', $version);
    }

    /**
     * A method to temporarily disable PEAR error handling by
     * pushing a null error handler onto the top of the stack.
     *
     * @static
     */
    function disableErrorHandling()
    {
        PEAR::pushErrorHandling(null);
    }

    /**
     * A method to re-enable PEAR error handling by popping
     * a null error handler off the top of the stack.
     *
     * @static
     */
    function enableErrorHandling()
    {
        // Ensure this method only acts when a null error handler exists
        $stack = &$GLOBALS['_PEAR_error_handler_stack'];
        list($mode, $options) = $stack[sizeof($stack) - 1];
        if (is_null($mode) && is_null($options)) {
            PEAR::popErrorHandling();
        }
    }

    /**
     * Returns the option from config or the default value if that option
     * do not exist.
     *
     * @param string $section  Section name
     * @param string $name     Option name
     * @param string $default  Default value to return if the option we
     *                         are looking for do not exist.
     * @return string
     */
    function getConfigOption($section, $name, $default = null)
    {
        if (isset($GLOBALS['_MAX']['CONF'][$section][$name])) {
            return $GLOBALS['_MAX']['CONF'][$section][$name];
        }
        return $default;
    }
}
###END_STRIP_DELIVERY

?>
